File: client-sdk/python/src/synadia_ai/agents/identity/signed_publish.py
"""Signing an arbitrary publish with a client's sender identity.

The core of :meth:`Agents.sign_sender` / :meth:`Agents.publish_signed` /
:meth:`Agents.request_signed`, shared with the signing handle a prompt
interceptor receives (:mod:`synadia_ai.agents.interceptor`), so a message
an interceptor publishes is signed exactly like one the application
publishes itself.
"""

from __future__ import annotations

from collections.abc import Mapping
from typing import TYPE_CHECKING

from ..errors import IdentityError, SenderSignatureRequiredError
from .options import Identity, plan_sender_header
from .sender_header import (
    AGENT_SENDER_HEADER,
    AgentSenderHeader,
    is_valid_sender_nonce,
    serialize_sender_header,
)

if TYPE_CHECKING:
    from nats.aio.client import Client as NATSClient

#: JetStream de-duplication header; ``publish_signed`` sets it to the nonce.
NATS_MSG_ID_HEADER = "Nats-Msg-Id"


async def signed_sender_header(
    identity: Identity | None,
    nc: NATSClient,
    subject: str,
    payload: bytes,
    *,
    sub: str | None = None,
    nonce: str | None = None,
) -> AgentSenderHeader:
    """The signed ``Agent-Sender`` header for a publish of ``payload`` to ``subject``.

    ``nonce``, when given, is the header's nonce in place of a fresh one —
    for a message whose body carries its own id. It must match
    ``[A-Za-z0-9_-]{1,64}`` and be unique per signer: a receiver refuses a
    nonce it has already seen from the same user.

    Raises :class:`SenderSignatureRequiredError` when no signer is
    configured, :class:`IdentityError` for a malformed ``nonce``, else the
    ``self_id()`` error when the identity is unavailable.
    """
    if identity is None or identity.signer is None:
        raise SenderSignatureRequiredError(subject)
    if nonce is not None and not is_valid_sender_nonce(nonce):
        raise IdentityError("nonce must match [A-Za-z0-9_-]{1,64}")
    plan = await plan_sender_header(
        identity, nc, sub if sub is not None else subject, require_signed=True
    )
    if plan is None:  # unreachable with a signer; keeps the type checker honest
        raise SenderSignatureRequiredError(subject)
    return await plan.build(payload, nonce=nonce)


async def signed_publish_headers(
    identity: Identity | None,
    nc: NATSClient,
    subject: str,
    payload: bytes,
    *,
    sub: str | None = None,
    headers: Mapping[str, str] | None = None,
    nonce: str | None = None,
) -> dict[str, str]:
    """:func:`signed_sender_header` merged into ``headers``, with its ``Nats-Msg-Id``.

    ``Agent-Sender`` and ``Nats-Msg-Id`` win over entries of the same name.
    """
    header = await signed_sender_header(identity, nc, subject, payload, sub=sub, nonce=nonce)
    hdrs: dict[str, str] = dict(headers) if headers else {}
    hdrs[AGENT_SENDER_HEADER] = serialize_sender_header(header)
    if header.nonce is not None:
        hdrs[NATS_MSG_ID_HEADER] = header.nonce
    return hdrs


def to_bytes(payload: bytes | str) -> bytes:
    return payload.encode("utf-8") if isinstance(payload, str) else payload


__all__ = [
    "NATS_MSG_ID_HEADER",
    "signed_publish_headers",
    "signed_sender_header",
    "to_bytes",
]

File: client-sdk/typescript/src/identity/signed-publish.ts
// Signing an arbitrary publish with a client's sender identity: the core of
// `Agents.signSender` / `publishSigned` / `requestSigned`, shared with the
// signing handle a prompt interceptor receives (`prompt/interceptor.ts`),
// so a message an interceptor publishes is signed exactly like one the
// application publishes itself.

import { headers, type MsgHdrs } from "@nats-io/nats-core";
import { IdentityError, SenderSignatureRequiredError } from "../errors.js";
import type { IdentityContext } from "./context.js";
import {
  AGENT_SENDER_HEADER,
  isValidSenderNonce,
  serializeSenderHeader,
  type AgentSenderHeader,
} from "./sender-header.js";

/** JetStream de-duplication header; `publishSigned` sets it to the nonce. */
export const NATS_MSG_ID_HEADER = "Nats-Msg-Id";

/** Options for the signed low-level wrappers. */
export interface SignedPublishOptions {
  /**
   * Subject to sign instead of the publish subject — only for a caller
   * whose own account renamed the import (sign the exporter's subject).
   */
  readonly sub?: string;
  /** Existing headers to add `Agent-Sender` (and `Nats-Msg-Id`) to. */
  readonly headers?: MsgHdrs;
  /**
   * The header's nonce, in place of a fresh one — for a message whose body
   * carries its own id: sign with that id and it is the nonce and the
   * `Nats-Msg-Id` as well, so a reader de-duplicating on `(user, nonce)`
   * and a stream de-duplicating on the message id see one message once.
   * Must match `[A-Za-z0-9_-]{1,64}` and be unique per signer: a receiver
   * refuses a nonce it has already seen from the same user.
   */
  readonly nonce?: string;
}

/**
 * The signed `Agent-Sender` header for a publish of `payload` to `subject`.
 * Rejects with {@link SenderSignatureRequiredError} when no signer is
 * configured, with {@link IdentityError} for a malformed `nonce`, else with
 * the `selfId()` error when the identity is unavailable.
 */
export async function signedSenderHeader(
  identity: IdentityContext | undefined,
  subject: string,
  payload: Uint8Array,
  opts: Pick<SignedPublishOptions, "sub" | "nonce">,
): Promise<AgentSenderHeader> {
  if (!identity?.signer) throw new SenderSignatureRequiredError(subject);
  const nonce = opts.nonce;
  if (nonce !== undefined && !isValidSenderNonce(nonce)) {
    throw new IdentityError("nonce must match [A-Za-z0-9_-]{1,64}");
  }
  const plan = await identity.plan(opts.sub ?? subject, true);
  if (!plan) throw new SenderSignatureRequiredError(subject); // unreachable with a signer
  return plan.build(payload, nonce);
}

/** {@link signedSenderHeader}, set on `opts.headers` (or fresh ones) with its `Nats-Msg-Id`. */
export async function signedPublishHeaders(
  identity: IdentityContext | undefined,
  subject: string,
  payload: Uint8Array,
  opts: SignedPublishOptions,
): Promise<MsgHdrs> {
  const h = await signedSenderHeader(identity, subject, payload, opts);
  const hdrs = opts.headers ?? headers();
  hdrs.set(AGENT_SENDER_HEADER, serializeSenderHeader(h));
  if (h.nonce !== undefined) hdrs.set(NATS_MSG_ID_HEADER, h.nonce);
  return hdrs;
}

export function toBytes(payload: Uint8Array | string): Uint8Array {
  return typeof payload === "string" ? new TextEncoder().encode(payload) : payload;
}
